Doc improvements (#695)
Azure/autorest.az: README.md

> # Contributing

Reviewer: Move this part to other part of this doc. I want to know what this project is when I open this doc.

Author: moved

This project welcomes contributions and suggestions.  Most contributions require you to agree to a
Contributor License Agreement (CLA) declaring that you have the right to, and actually do, grant us
the rights to use your contribution. For details, visit https://cla.opensource.microsoft.com.

When you submit a pull request, a CLA bot will automatically determine whether you need to provide
a CLA and decorate the PR appropriately (e.g., status check, comment). Simply follow the instructions
provided by the bot. You will only need to do this once across all repos using our CLA.

This project has adopted the [Microsoft Open Source Code of Conduct](https://opensource.microsoft.com/codeofconduct/).

> For more information see the [Code of Conduct FAQ](https://opensource.microsoft.com/codeofconduct/faq/) or

Reviewer 2: Just reference CODE_OF_CONDUCT.md

Author: removed.

contact [opencode@microsoft.com](mailto:opencode@microsoft.com) with any additional questions or comments.

# Introduction

> This project is for Azure CLI Code Generator. We support to generate as both cli extensions and cli main repo modules. In this README file, we will briefly introduce how the Azure CLI Code Generator works. Secondly, we will introduce how to use the CLI code generator and what kinds of advanced features of CLI code generator do we support. Then, we will introduce how to debug the CLI Code Generator. Finally, we will end this document with Autorest Pipeline definitions.  

Reviewer 3: could we add a catalog with hyperlink in this part?
here is an example in cli repo https://github.com/Azure/azure-cli/blob/dev/doc/authoring_command_modules/authoring_commands.md

Author: added. still trying to make a better layout.

Author: Can you check again? Thanks

> This project is for Azure CLI Code Generator. We support to generate as both cli extensions and cli main repo modules. In this README file, we will briefly introduce how the Azure CLI Code Generator works. Secondly, we will introduce how to use the CLI code generator and what kinds of advanced features of CLI code generator do we support. Then, we will introduce how to debug the CLI Code Generator. Finally, we will end this document with Autorest Pipeline definitions.  

Reviewer: ```suggestion
This project is for Azure CLI Code Generator. It supports both Azure CLI extensions and Azure CLI main repository modules generation. This document introduces how Azure CLI Code Generator works, how to use it, advanced features. Then, it introduces how to debug the code generator. Finally, it gives Autorest Pipeline definitions.  
```

Author: changed

> This project is for Azure CLI Code Generator. We support to generate as both cli extensions and cli main repo modules. In this README file, we will briefly introduce how the Azure CLI Code Generator works. Secondly, we will introduce how to use the CLI code generator and what kinds of advanced features of CLI code generator do we support. Then, we will introduce how to debug the CLI Code Generator. Finally, we will end this document with Autorest Pipeline definitions.  

Reviewer 2: Not sure what this mean..  'We support to generate as both cli extensions and cli main repo modules.'

Author: Changed.

# Assumption

This document is mainly for Azure CLI developers and people who are interested in generating Azure CLI by themselves.  
# How Does the CLI Code Generator Work.

> The Azure Code Generator is basically a [Autorest](https://github.com/Azure/autorest) extension which takes swagger in [azure-rest-api-specs](https://github.com/Azure/azure-rest-api-specs) or [azure-rest-api-specs-pr](https://github.com/Azure/azure-rest-api-specs-pr) repos as input and output functional Azure CLI code. It uses Autorest V3 to handle the configuration interpretation, pipeline resolving, pipeline scheduling and uses [Autorest.Modelerfour](https://github.com/Azure/autorest.modelerfour) as a basic code model. 

Author: Yes, but it's actually not necessary if you can provide all the contents in command line. For example, users can also use `autorest --az --input-file=swagger.json --az.extensions=extensionName --az.namespace=azure.mgmt.rp --az.package-name=azure-mgmt-rp --az-output-folder=$(azure-cli-extension-folder)/src/rp --python-sdk-output-folder=$(az-output-folder)/azext_rp/vendored_sdks/rp --azure-cli-extension-folder=../azure-cli-extenions`  to generate the code. In such case, no readme file is needed.

Both the Autorest.Clicommon and Autorest.Python take Autorest.Modelerfour as input, Normally autorest.python will not flatten the code model because the Python Language naturally support complex object, they don't need to flatten the parameter but for command line tools like Azure CLI, the parameter layer must not deeper than 2, otherwise it's hard to express the complex object without using a bunch of delimiters. Therefore, autorest.clicommon will flatten the code model. Autorest az will merge the two code model from autorest.python and autorest.clicommon.    

After autorest.az has done some CLI specific logic, it will render the code model into code template finally output the generated code.   

# How to use the CLI Code Generator

Users can use autorest.az to generate the Azure CLI code both in rest api specs PR pipeline and in their local environment. Guidance for generate the Azure CLI code in rest api specs PR pipeline can be found in [rest api specs documentation](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/code-gen/configure-cli.md). Here we only focus on the guidance to generate the CLI code in local.  

> More detailed generation guidance can be found in the [how to generate doc](https://github.com/Azure/autorest.az/blob/master/doc/how-to-generate.md)

Author: I want this doc to give users a general impression of how to use the CLI codegen, and if users want to know more, they can open the linked doc to find out the details.

> ## Preparing Environment

Author: Yes. In this doc we only cover the local run autorest way, the swagger pipeline and doc is not covered here. But I give the link in above.

In order to generate the Azure CLI, we need 
1. Nodejs v10+  

> better to be v12+

Reviewer 3: could we provide nodejs installation guide with something like below?
`to install NodeJS. You could install a NodeJS globally or use nvm (for linux) or nvm-windows (for windows). It will also help to install NodeJS package management command line tool, npm.
`

Author: added

2. Python3  
we need a Python venv  
```
python -m venv env
source ./env/bin/activate // or .\env\Scripts\Activate.ps1 in windows
pip install azure-cli // This is must to have if for simple try out
pip install azdev // this is optional if for simple try out. 
```
3. Autorest  
it can be installed by running `npm install -g autorest@latest` 
4. Repos (Optional)  

> if users only want to try out the generated CLI extensions, they can also build the wheel artifact by themselves with `python setup.py sdist bdist_wheel` in the azure-cli-extensions/src/{serviceName} folder.  

Reviewer 3: one minor suggestion: could we use second person  `you` not `they` to let users feel like you are sharing with the reader?

Author: changed.

* Users need to prepare their swagger and prepare  
[azure-cli-extensions](https://github.com/Azure/azure-cli-extensions) repo if targeting at Azure CLI extensions development. Or the generated code needs some manual customization after trying out.   
[azure-cli](https://github.com/Azure/azure-cli) repo if targeting at Azure CLI main repo modules development.   

> 

Reviewer 3: should we also add swagger repo here?

Author: Added. I previously hesitate about this because for some new RP their swagger probably hasn't been merged into the swagger repo yet.

## Authoring readme files
Because the Autorest.Az depends on Autorest.Clicommon and Autorest.Python, it needs three readme files the readme.az.md, readme.cli.md and readme.python.md to be ready before we use it.

Users can refer to [this document](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/code-gen/configure-cli.md) for more details. 

> ## Generating the code

Reviewer: ```suggestion
## Generating code
```

Author: The Code means the generated code. Don't think we should remove 'the'.

1. A simple command for authoring Azure CLI extensions would be  

> `autorest --az --use=@autorest/az@latest <path-to-the-swagger-readme.md> --azure-cli-extension-folder=<path-to-the-azure-cli-extension-repo>`

Reviewer 3: recommend to add all recommended options in this sample. For example, could we add `--sdk-no-flatten`  in the example because it would align with official track2 SDK format

Author: Actually the other command line options are lin the  https://github.com/Azure/autorest.az/pull/695/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R102 linked doc. Maybe I can add a general intro here.

Author: Added a section of Different Generation Options https://github.com/Azure/autorest.az/pull/695/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R59 section and updated the autorest command samples.

2. A simple command for authoring Azure CLI main modules would be  
`autorest --az --use=@autorest/az@latest <path-to-the-swagger-readme.md> --azure-cli-folder=<path-to-the-azure-cli-repo>`

> The `--use=@autorest/az@latest` will use the latest `@autorest/az` public releases, private releases can be found in [github releases](https://github.com/Azure/autorest.az/releases) if users want to try out some private releases for preview some new features and fixes.  

Reviewer 3: could we also add sample for use private build package with the download url?

Author: added. Thanks

Other command line options can be found in the [generate with different options doc](https://github.com/Azure/autorest.az/blob/master/doc/how-to-generate-with-different-options.md)

## Building the code
If it's try out mode, 
``` 
cd <az-output-folder> // the az-output-folder you have specified in your readme.az.md 
python setup.py sdist bdist_wheel // you will find a wheel file in your local dist folder
az extension add <path-to-the-wheel-file>
``` 
If it's for azure-cli-extensions development, you need  
```
azdev setup -r ./azure-cli-extensions -c ./azure-cli 
// where the ./azure-cli-extensions is the path to azure-cli-extensions repo and ./azure-cli is the path to azure-cli repo, you don't need to specify them both if you are not working on both of them
azdev extension add <extension-name> // for Azure CLI extensions
// extension add step is needed for developing Azure CLI main repo modules 
```
## Play with Azure CLI

> Now you can run `az <extension-name> -h` to see all the commands and parameters. You can also find a report.md in generated azext_{extensionName} folder, which contains an overview of all the generated command groups, commands and parameters.   

Reviewer 3: We may have the scenario that the extension name is different from main command group name?

Author: Yes. I changed the doc. Just thought if it may be confusion for users if we say main command group name. In the Folder Customization, I mentioned each place holder can be changable.

      set:
          group: datashare consumer trigger

    ## change the group and name of a command
    - where:
          command: datafactory integration-runtime create-linked-integration-runtime
      set:
          command: datafactory integration-runtime linked-integration-runtime create

```
See [how to add or remove subgroups](https://github.com/Azure/autorest.az/blob/master/doc/faq.md#how-to-addremove-subgroup) for more details.   

### **rename/hide command groups, commands, parameters**  
We provide the ability for user to rename or hide command groups or commands or parameters. For example:
```
cli:
  cli-directive:
    ## rename a parameter 
    - where:
        group: groupCondition
        op: opCondition
        param: paramCondition
      name: new_op_name

    ## hide an operation
    - where:
        group: groupCondition
        op: opCondition
      hide: true    
```
* Note: if a parameter has the flattened schema prefix in the name, then we can't rename it in this way, because in Autorest.Clicommon it doesn't have the flattened schema prefix. We can only add alias for this parameter in such case.

### **Client Factory Customization**

> By default, the client factory should not expose the subscriptionId, baseUrl parameters to customer as well as authentication policy but we provide customization for these scenarios.

Reviewer 4: for each of the advanced feature, is it ok to add a real RP as example, so people could check RP behavior for better understandign?

Author: Do you mean the generated code example? I have provide the readme.az.md configuration example for most of them.  And I was thinking maybe we can add a user scanrios folder and CLI guys can add some real user scenarios while onboarding RP.
